### PR TITLE
Add apply_cmt_version function and xenonnt context

### PR DESCRIPTION
**Pull request by ershockley**
## What does the code in this PR do / what does it improve?
It makes it easy to set the CMT global version for a context. It also adds a new context function that just takes the cmt version as an argument.

## Can you briefly describe how it works?
The standard xenonnt_online context is initialized as normal, and then the config is updated according to the global CMT version passed as an argument. 

## Can you give a minimal working example (or illustrate with a figure)?
```
import straxen

st1 = straxen.contexts.xenonnt()
hash1 = st1.provided_dtypes()['peaklets']['hash']
st2 = straxen.contexts.xenonnt('v1')
hash2 = st2.provided_dtypes()['peaklets']['hash']
print(hash1, hash2)

# returns
xz7qdkrg3p o37giwozd6
```

### CODE REVIEW

**Comment by ershockley** [comment]
I don't think i've got all the corrections here yet -- @ahiguera-mx could you maybe help me make sure we have them all covered?

**Comment by ahiguera-mx** [comment]
Hi @ershockley

I think is better to use the new convention for the CMT option based on PR 457(https://github.com/XENONnT/straxen/pull/457), otherwise we would have to modify them again once PR457 is merged so I think we should wait, in addition, instead of hard coding these we should have them in the DB (I'm working on that) so we can return all correction options with a single function

**Comment by ershockley** [comment]
Yes I figured we would change it after #457 is merged (which is hopefully soon?). What do you mean by putting them in the DB instead of hardcoding though?

**Comment by ahiguera-mx** [comment]
Instead of hardcoding the options in `apply_cmt_version()` we save them in corrections DB and make this function to query the DB and return the dictionary, in that sense if we add a new correction we don't have to add it without modifying the code

**Comment by ershockley** [comment]
If we add a new correction/option we would have to implement it here in straxen anyway though, right? Also a DB query might not be compatible with specific straxen versions after e.g. #457 where the conventions change. Personally I feel like hardcoding here would be better but maybe I'm missing something.

**Comment by ahiguera-mx** [comment]
you are right, if we add a new correction we would have to implemented the new strax.option in straxen at the context level...

File: straxen/corrections_services.py [hunk 1]
```diff
@@ -254,3 +254,21 @@ def cacheable_naming(*args, fmt='.npy', base='./resource_cache/'):
 class GainsNotFoundError(Exception):
     """Fatal error if a None value is returned by the corrections"""
     pass
+
+
+@strax.Context.add_method
+def apply_cmt_version(context, cmt_version):
```
Comment (WenzDaniel) [inline straxen/corrections_services.py:260]: Although the shape of the options makes it already quite clear that these are CMT options, we could in principle agree on some semantics here. E.g. we could agree that all cmt options should start with a `cmt_`. This would allow you to loop over them more easily.

**Review by jorana — CHANGES_REQUESTED** [review]
Hi Evan and Aaron. Thanks I think we indeed desperately need this.

However, we should be implementing a global version rather than applying e.g. v1 for all the options. So we should make sure that global version are referring to the straxen options, it why we have the global versions in CMT in the first place. For example, we might do a SR with hit-finding thresholds on the ONLINE version while an FDC is at v2.

If the global version is incompatible with straxen options, I think we should fix that.

File: straxen/corrections_services.py (hunk 1 above)
Comment (jorana) [reply straxen/corrections_services.py:260]: Very good idea, I like it

Comment (ershockley) [reply straxen/corrections_services.py:260]: I'm not opposed to this in principle, but wouldn't it have been more relevant for #457? Also these options don't always go through CMT right? Though I guess the vast majority of time they will.

**Comment by ershockley** [comment]
My understanding is that the version shown here is indeed the global version, but of course I could be wrong as I often am :) 

For your example, there would just be a global version where the thresholds were at _local version_ ONLINE/v0 and FDC map v2. That global version is what gets passed here.

**Comment by ahiguera-mx** [comment]
You are right Evan, versions are global version, and you example is correct

**Comment by ahiguera-mx** [comment]
Hi guys
As discussed I've implemented some tweaks, the overall idea is to set all the strax.options config that use CMT with a global version (this makes super easy the bookkeeping). However, under the hood `apply_cmt_version()` sets local version for every strax.option() in that sense we don't t change the lineage for things that do not change even if we uses a new global version.

File: straxen/corrections_services.py [hunk 2]
```diff
@@ -98,34 +98,65 @@ def _get_correction(self, run_id, correction, global_version):
...
+                    df_global = self.interface.read('global' if self.is_nt else 'global_xenon1t')
+                    gains = df_global['ONLINE'][0]
+                    pmts = list(gains.keys())
+                    for it_correction in pmts: # loop over all PMTs
+                        if correction in it_correction:
+                            df = self.interface.read(it_correction)
+                            df = self.interface.interpolate(df, when)
+                            values.append(df.loc[df.index == when, version].values[0])
+                else:
+                    df = self.interface.read(correction)
```
Comment (WenzDaniel) [inline straxen/corrections_services.py:123]: Hej guys, I started looking into the latest changes related to the local version. I think in general the code can be simpler and less nested. Before I make some detailed suggestion I have a question though. I am a bit confused about the PMT part as we also have the function `get_pmt_gains`. What is the difference, what are the plans here? 
Further, I am wondering if a general change on how we query the corrections would be beneficial. Currently, we do local_version -> global_version, query with global version. But I think this is unnecessarily complicated. In the future we will always query via the local_versions (which feels much more natural to me). Should not we reverse the logical structure:
* local_version -> query local_version
* global_version -> convert to local versions -> query with local_versions.
In that case we can move the "convert to local versions" part in a separate function making the code much cleaner and more readable. Are there any technical constraints preventing us from doing this?

Comment (ahiguera-mx) [reply straxen/corrections_services.py:123]: Hi Daniel,

Thanks for the useful comments, I'm pretty sure that the code can be improved, regarding your comment about local and global I have a few comments.
We always query local version at the end, global is just a map to a set of local versions, the way I implemented allows the use of either use global version or local version, I think we want to have both options. Specially for bookkeeping with global version we(production) don't have to worry about every single local version.
If we only query local version then we will have to build that map to get the set of global versions, so I don't see the advantage of doing that. 
PMTs are special case since every single PMT is its own dataframe wich complicates things

Comment (ahiguera-mx) [reply straxen/corrections_services.py:123]: I have included a few examples regarding the uses of local and global at the end of the notebook [here](https://github.com/XENONnT/corrections/blob/master/CMT/examples/CMT_user_guide.ipynb) if you want to check it

**Comment by ershockley** [comment]
Thanks @ahiguera-mx. I tend to agree with Daniel and would prefer we just move to using local_version within most of CMT and then have a simple function that converts from global --> local that then goes inside `apply_global_version`. I thought that was what we discussed last week maybe we weren't quite on the same page. Do you see a problem with doing it this way? 

I will play around a bit and maybe propose something here. I should learn more about the internals of CMT anyway.

**Comment by ahiguera-mx** [comment]
I'm not sure what is the issue here, the only thing is how we access the information, either via global version or local version, the current implement allows to use both, at then end CMT always query local version. Also we need global to make bookkeeping easy, so production just need to set a global version vX and CMT does the rest... regardless how access the information. 
Maybe I don't understand what you are saying so it probably better if you can draft something to better understand the issue

**Comment by ahiguera-mx** [comment]
Hi guys,

After a very useful discussion with Evan, I made the corresponding changes, now any strax.Option() only will use a local version from CMT. As discussed this will avoid re-processing since very often local version won't change for the low-level data. 
To do bookkeeping and make life easy for production we use the concept of global_version, in that case `apply_cmt_version(context, cmt_global_version)` will get a global_version and will look at the local_version for every single correction. The convention then is going to be local_version = vx, global_version = global_vx. 
There is one special case that is the PMT that need to be grouped since every pmt is its own dataframe, in the past I did not need to do that since all PMTs were grouped in the global version, since that is not the case anymore need to hack that part.

Thanks for all very useful comments

**Comment by ershockley** [comment]
Ok from my perspective this PR should be good to go now.  @jorana @WenzDaniel  are the changes you requested implemented? 

Also @ahiguera-mx I made a few minor changes if you want to take a look.

**Comment by WenzDaniel** [comment]
Hi @ershockley I am sorry, but either I am doing something wrong or there are still some bugs in your code. The following things I noticed:

If I try to use `st = straxen.contexts.xenonnt()` I get the following error:
```python
ValueError: Global version ONLINE not found! Try one of these:
		global_ONLINE
		global_v1
		global_v2
		global_v3
```
So here you need to change the default. However, online the global version `global_ONLINE` is currently working. All the other versions return:
```python
/mymodules/straxen/straxen/corrections_services.py in apply_cmt_version(context, cmt_global_version, posrec_algo)
    322                       cnn_model=("cnn_model", local_versions['cnn_model'], True),
    323                       electron_drift_velocity=("electron_drift_velocity", local_versions['electron_drift_velocity'], True),
--> 324                       electron_drift_time_gate=("electron_drift_time_gate", local_versions['electron_drift_time_gate'], True),
    325                       baseline_samples_nv=('baseline_samples_nv', local_versions['baseline_samples_nv'], True),
    326                       hit_min_amplitude=('hit_thresholds_tpc', local_versions['hit_thresholds_tpc'], True),

KeyError: 'electron_drift_time_gate'
```
Further, I think it is going to be a bit confusing for the analysts that `st = straxen.contexts.xenonnt('global_v1')`  takes as a first argument the CMT version instead of a directory now. I think we can save us here a lot of complains that "things are not working" if we keep as a first argument `output_folder='./strax_data'`.

Last but not least, if xenonnt should become the default context we also should keep in mind to updated cutax accordingly.

File: straxen/get_corrections.py [hunk 3]
```diff
@@ -118,6 +119,22 @@ def is_cmt_option(config):
...
+    with option name as key and default values as values.
+
+    :param context: Context with registered plugins.
+    """
+    default_cmt_options = {}
+    for data_type, plugin in context._plugin_class_registry.items():
+        for option_key, option in plugin.takes_config.items():
+            if straxen.get_corrections.is_cmt_option(option.default):
+                default_cmt_options[option_key] = option.default
+    return default_cmt_options
```
Comment (jorana) [inline straxen/get_corrections.py:135]:
```suggestion
def get_cmt_options(context):
    """
    Function which loops over all plugin configs and returns dictionary
    with option name as key and current settings as values.

    :param context: Context with registered plugins.
    """
    cmt_options = {}
    for data_type, plugin in context._plugin_class_registry.items():
        for option_key, option in plugin.takes_config.items():
            if (option_key in context.config and
                straxen.get_corrections.is_cmt_option(context.config[option_key])
                ):
                cmt_options[option_key] = context.config[option_key]
            elif straxen.get_corrections.is_cmt_option(option.default):
                cmt_options[option_key] = option.default

    return cmt_options
```    

This would also get the overwritten options properly
```

**Review by jorana — CHANGES_REQUESTED** [review]
Thanks guys for all the hard work 😄 

Generally, I think there are some good points raised by Daniel, the idea of checking the  context and working from there sounds great. There is a catch for the mlp-derived options though, for these I think it would not work right @WenzDaniel ?

I think there is still some stuff we need to do on the CMT side, e.g. specifying v1 does not work, it looks like some options are not in CMT v1 (see the failing test). Should we set that to ONLINE if nothing is returned by CMT?

Also, would you agree this might be a great time to add a graph or something to the documentation since this PR completely changes the way CMT works and to many people it's a black box?

**Comment by WenzDaniel** [comment]
Thanks for the addition to the function. I do not fully understand what you mean with: 

> There is a catch for the mlp-derived options though, for these I think it would not work right.

Should not the corrections for the different position reconstruction algorithms be handled here: https://github.com/XENONnT/straxen/blob/c23664c274b873c72cbceab7d51abe3917e6ba83/straxen/plugins/event_processing.py#L387-L389

File: straxen/corrections_services.py [hunk 4]
```diff
@@ -255,3 +297,36 @@ def cacheable_naming(*args, fmt='.npy', base='./resource_cache/'):
...
+                      mlp_model=("mlp_model", local_versions['mlp_model'], True),
+                      gcn_model=("gcn_model", local_versions['gcn_model'], True),
+                      cnn_model=("cnn_model", local_versions['cnn_model'], True),
+                      electron_drift_velocity=("electron_drift_velocity", local_versions['electron_drift_velocity'], True),
+                      electron_drift_time_gate=("electron_drift_time_gate", local_versions['electron_drift_time_gate'], True),
+                      baseline_samples_nv=('baseline_samples_nv', local_versions['baseline_samples_nv'], True),
+                      hit_min_amplitude=('hit_thresholds_tpc', local_versions['hit_thresholds_tpc'], True),
+                      hit_min_amplitude_he=('hit_thresholds_he', local_versions['hit_thresholds_he'], True),
+                      hit_min_amplitude_nv=('hit_thresholds_nv', local_versions['hit_thresholds_nv'], True),
+                      hit_min_amplitude_mv=('hit_thresholds_mv', local_versions['hit_thresholds_mv'], True),
```
[the remark this answers is not in the document]
Comment (ershockley) [reply straxen/corrections_services.py]: @WenzDaniel I like the idea of this function. My only concern is that it relies on the options to have already been in CMT format before applying the function. Is this a safe assumption?

File: straxen/corrections_services.py [hunk 5]
```diff
@@ -77,66 +79,76 @@ def get_corrections_config(self, run_id, config_model=None):
         """
 
         if not isinstance(config_model, (tuple, list)) or len(config_model) != 2:
-            raise ValueError(f'config_model {config_model} must be a tuple')
-        model_type, global_version = config_model
+            raise ValueError(f'config_model {config_model} must be a tuple of length 2')
+        model_type, version = config_model
```
[the remark this answers is not in the document]
Comment (ershockley) [reply straxen/corrections_services.py:85]: This will not work as the arguments are different right? The config passed to is_cmt_option is a tuple of length 3 (includes the True/False for nT) and config passed here is length 2.

**Comment by ershockley** [comment]
Thanks for the suggestions/changes guys. Just to comment quickly on a few of them.

On the different global versions, I have to say I only tested two of them (online and most recent one) which do work (others don't due to the electron_drift_time_gate key error) 😞. I'm not sure what the best thing to do here is. If the correction was not in CMT at the time we fixed that global version then I don't know if we should be adding in an online version of a new correction after the fact. This is in a way changing the fixed versioning that we wanted right? I think I would prefer to raise an error somewhere that this version is not compatible with the current straxen version. I'm not sure though -- it's probably easier to just grab the online version, but I worry a bit about doing things like this. 

>Further, I think it is going to be a bit confusing for the analysts that st = straxen.contexts.xenonnt('global_v1') takes as a first argument the CMT version instead of a directory now. I think we can save us here a lot of complains that "things are not working" if we keep as a first argument output_folder='./strax_data'.
I disagree. The cmt version is going to effectively become part of the strax(en) version that every analyst will need to report for their work, and will of course change the lineage hashes so needs to be one of the first things people think about when setting up the context. They can pass local options like where they store local data after that. There might be some issues at the beginning as you say but I think this is better in the long run.

I will fix some of those issues later today -- indeed not checking the default xenonnt context is quite embarrassing 😊

File: straxen/corrections_services.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (WenzDaniel) [reply straxen/corrections_services.py]: I think so yes. I doubt that people will forget to change the setting after moving a correction to CMT. However, I think it is more likely that they will forget to update the hardcoded list above which will make a mess.

[the remark this answers is not in the document]
Comment (ershockley) [reply straxen/corrections_services.py]: Yeah I see your point, but this does not just depend on people updating default options right? While I don't see this happening much, one can imagine modifying a context in some way and then after that calling `apply_cmt_version`, where the changes that person made could affect how the version gets applied. It seems safer to me to just hardcode all the known corrections used for a given straxen version. I think the corrections experts/team leaders/ACs should be aware of this important function and make sure it gets updated if there is a new correction added (which hopefully doesn't happen much anyway). I would prefer this over assuming a certain convention that's not strictly enforced. 

 Or -- ideally -- sync the naming scheme between CMT and the strax plugins. But not sure this is doable now.

[the remark this answers is not in the document]
Comment (WenzDaniel) [reply straxen/corrections_services.py]: >  would prefer this over assuming a certain convention that's not strictly enforced.
But it is is not it? 
If you specify e.g. `'baseline_samples_nv': 18` cmt will complain.

[the remark this answers is not in the document]
Comment (WenzDaniel) [reply straxen/corrections_services.py]: > While I don't see this happening much, one can imagine modifying a context in some way and then after that calling apply_cmt_version, where the changes that person made could affect how the version gets applied. 

Could you give an example for such a modification? Why would we want to make a modification in xenonnt_online which gets overwritten in xenonnt? This does not make sense to me.

[the remark this answers is not in the document]
Comment (ershockley) [reply straxen/corrections_services.py]: I was under the impression you don't need to use CMT and that you can pass other options (like e.g. those old 1T maps stored in pax)?

[the remark this answers is not in the document]
Comment (ershockley) [reply straxen/corrections_services.py]: >Could you give an example for such a modification? Why would we want to make a modification in xenonnt_online which gets overwritten in xenonnt? This does not make sense to me.

I mean this is one of the main reasons we use strax to begin with right? The fact people can update configuration in their own context and reprocess quickly. Maybe someone is testing a new correction map and wants to switch between a new one not yet in CMT and an old one that is in CMT. 

So something like this (pseudocode)

```
st = xenonnt_online()
st.set_config('correction_map', 'some_new_correction_file')

# do something

# now want to apply global versions
st.apply_cmt_version('global_v1')
```
The apply_cmt_version in this case behaves differently depending on the state of the context passed to it. I think this is not ideal.

**Comment by ahiguera-mx** [comment]
> On the different global versions, I have to say I only tested two of them (online and most recent one) which do work (others don't due to the electron_drift_time_gate key error) 😞. I'm not sure what the best thing to do here is. If the correction was not in CMT at the time we fixed that global version then I don't know if we should be adding in an online version of a new correction after the fact. This is in a way changing the fixed versioning that we wanted right? I think I would prefer to raise an error somewhere that this version is not compatible with the current straxen version. I'm not sure though -- it's probably easier to just grab the online version, but I worry a bit about doing things like this.

Yes, old global_version would not have a complete list of corrections as in the latest global_version, I think the best solution is raise an error in `apply_cmt_version()`

File: tests/test_contexts.py [hunk 6]
```diff
@@ -26,6 +26,14 @@ def test_xenonnt_online():
     st.search_field('time')
 
 
+def test_xennonnt(cmt_version='ONLINE'):
+    st = xenonnt(_database_init=False, use_rucio=False)
+    st.search_field('time')
+
+def test_xennonnt_v1(cmt_version = 'v1'):
```
Comment (ahiguera-mx) [inline tests/test_contexts.py:33]: Yes, this won't work because old_global version have not a complete set of corrections as the latest one, we can check if global_v1 contains all corrections, if not, do not set the "local_version" and let the default value

File: straxen/corrections_services.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ershockley) [reply straxen/corrections_services.py]: I'm specifically concerned about someone calling `apply_cmt_version` outside of how the  `xenonnt` context gets initialized. It's an edge case so maybe not worth the time we are discussing here, but my view is that if we see a possible edge case issue that we should address it.

I also just don't think we should be assuming a convention in something as important as our corrections. I will propose an update to your function that doesn't make this assumption.

[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/corrections_services.py]: edge cases are a "feature" of strax(en) not a "bug"... this is where I would say strax(en) is basically anarchy

[the remark this answers is not in the document]
Comment (jorana) [reply straxen/corrections_services.py]: @ershockley , would making this function be not a context attribute an idea? Such that `st.apply_cmt_version` does not exist?

[the remark this answers is not in the document]
Comment (WenzDaniel) [reply straxen/corrections_services.py]: > edge cases are a "feature" of strax(en) not a "bug"... this is where I would say strax(en) is basically anarchy

lol wtf no idea what to reply to this. Maybe "Write your own tool?" This is really a very unappropriated comment.

[the remark this answers is not in the document]
Comment (ershockley) [reply straxen/corrections_services.py]: Come on guys, no need for this. Let's end this conversation here. 

@jorana this could be an option as well. Let me try a few things and come back here later today (in a new discussion thread).

[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/corrections_services.py]: @WenzDaniel I Did not mean to be disrespectful, my apologies if I gave the wrong impression... Let me rephrase it, since strax(en) allows user change any configs this potential leads to undesired challenges

**Comment by jorana** [comment]
@ershockley @ahiguera-mx , I just had a small discussion with Evan in gather town to push a quick commit on where/how to add some docs on CMT in 1ac31f7 , I hope Aaron is willing to help writing some on how CMT <-> straxen and especially the user is interfacing with it.

I hope the commit helps 😄

To build locally if you want to have a preview (modulo having to install some additional packages)
```bash
conda activate strax
cd streaxen/docs
conda install pandoc
pip install -r ../extra_requirements/requirements-docs.txt
bash make_docs.sh
# Results are written to docs/build
```

**Comment by ahiguera-mx** [comment]
@jorana
Sure, I will add the corresponding info regarding CMT

**Comment by ershockley** [comment]
@WenzDaniel I went ahead and changed things to use the nice function you wrote -- thanks for providing that! I'm still a bit concerned that it's not as clear as before but I think it will be fine for the reasons you say. I had to make a few hacks for the different gain models and for the maps that depend on different posrec algorithms, but I think it should be working now. One benefit I didn't realize before from using your function is that we don't need to pass a posrec algorithm anymore to apply_cmt_version which is nice (although parts of this I find a bit confusing... see below). 

@WenzDaniel @jorana @ahiguera-mx @l-althueser @jingqiangye (just tagging all relevant people) 
Relating to the posrec algorithms, I didn't realize that we use the variable for default position reconstruction when we get corrections, like [here](https://github.com/XENONnT/straxen/blob/2544f3cc44e2eb8547a83d741dfdd310859d33ae/straxen/plugins/event_processing.py#L388). Maybe it's just my non-expertise coming in here but I found this quite non-intuitive. In my view I shouldn't have to check 2 variables to see what specific correction map I am using in the context. Combing this with how the gain models are implemented I really think the interface between the 'strax' side of straxen and the 'cmt' side should be streamlined to communicate better. Why can't the CMT correction names be exactly the same as those used by straxen or vice versa? But let's discuss that more later, we really need to merge this soon. Just wanted to mention it here. 

@jorana I was running into issues using the database on the test suite -- do you have any idea what's going on there? I just disabled the tests for now. It looks like the utilix file isn't being created properly, but I didn't have time to look deeper. 

I think all the issues should now be addressed apart from the testing. But please let me know.

**Comment by jorana** [comment]
@ershockley I can fix the tests, thanks!

>Relating to the posrec algorithms, I didn't realize that we use the variable for default position reconstruction when we get corrections, like here. Maybe it's just my non-expertise coming in here but I found this quite non-intuitive. In my view I shouldn't have to check 2 variables to see what specific correction map I am using in the context.

@l-althueser @jingqiangye @reallsx @terliuk  One thing that is mostly due to some of my ideas is the FDC depending on the MLP algorithm. We discussed extensively in team-C and this appeared to be the preferred solution (set the default once such that users don't have to manually pick the MLP/GCN/CNN algo for x_mlp at event level while still allowing to compare the different positions of x_mlp, x_gcn etc.).

In case you/people think this is too confusing or too much work to maintain as such, there are two alternatives:
- We can of course do the dumb thing and just have no "S2_x" fields anywhere and only "S2_x_mlp" etc. fields. At the time, it was considered not so urgent to have always three variables for the different algos. @ftoschi made a nice [tool to compare](https://github.com/XENONnT/straxen/blob/master/straxen/analyses/posrec_comparison.py).
- Make it the analysts responsibly to specify both `fdc_map = (fdc_map, ONLINE, True) ` and `default_reconstruction_algorithm = mlp`. In my opinion this is very error prone.

**Comment by l-althueser** [comment]
To my mind it is still the preferred option to use the default position reconstruction algorithm of the context for applying corrections. I don't think that giving the user `cS1_<algo>` will be of any help since one would then need to look up which NN one should use. Having a simple `cS1` is much cleaner and if we find that one NN is the best we can just adjust the context and be sure that everybody follows the best approach.
We kept `z` in the variables for a similar reason (which is currently a copy of `z_naive`) and also have `x`,`y`. The reason we still have the positions for all three NNs in one context is simply the possibility for cuts.

Just noticed that I repeated Jorans words .. sorry but I still like the current approach. We could maybe try to think about a function which resolves all the names and just gives the file names that are actually used (probably not easy to implement).

**Review by jorana — APPROVED** [review]

**Comment by ershockley** [comment]
Thanks for the comments all! Just to respond to one last thing: 

>Make it the analysts responsibly to specify both fdc_map = (fdc_map, ONLINE, True) and default_reconstruction_algorithm = mlp. In my opinion this is very error prone.

This is exactly how it's implemented right now, no? At least that is my understanding of [this line](https://github.com/XENONnT/straxen/blob/db4926a2dbdf272c4d07ae352625e5d897f7b9da/straxen/plugins/event_processing.py#L388). We pass the version of the FDC map (though it's unclear which map it is at the time) via the CMT option, but the actual map comes from the default_reconstruction_algorithm. So one needs to check both the FDC map version and this default_reconstruction_algorithm option to get what the FDC map is (same for S1 LCE, I think). This is the part I find very confusing and as you say error prone. 

I am totally happy with how the posrec variables are handled -- I'm more concerned about how the configuration is set. 

Anyway I think we can merge this now. Thanks all for the discussion and to @ahiguera-mx @jorana @WenzDaniel for your contributions!

**Comment by jorana** [comment]
Just for completeness, I was perhaps not clear. The situation I was trying to explain is that otherwise one could do this:
-  fdc_map = (fdc_map_**GCN**, ONLINE, True) and default_reconstruction_algorithm = **mlp**

Which is what I would call error prone

**Comment by l-althueser** [comment]
I would hope that analysts just work with local versions when changing maps. This would prevent errors like that but would lead to even more difficult to catch issues. Most corrections are based on each other and there actually is only one possible combination that makes sense to use (which is the one from the global version). Otherwise one would for example use a S1 map made of a different FDC map and similar.
So our general advice to analysts is to not change any of the options except if they are experts for corrections and want to test a specific case (e.g. they update their maps). I see that there is much potential to set completely wrong configs but also that there is no way to really validate any combination that one can enter. We could throw warnings ..
